src/libstd/sys/wasm/thread.rs

```diff
@@ -59,48 +59,32 @@ pub mod guard {
     pub unsafe fn init() -> Option<Guard> { None }
 }
 
-cfg_if::cfg_if! {
-    if #[cfg(all(target_feature = "atomics", feature = "wasm-bindgen-threads"))] {
-        #[link(wasm_import_module = "__wbindgen_thread_xform__")]
-        extern {
-            fn __wbindgen_current_id() -> u32;
-            fn __wbindgen_tcb_get() -> u32;
-            fn __wbindgen_tcb_set(ptr: u32);
+// This is only used by atomics primitives when the `atomics` feature is
+// enabled. In that mode we currently just use our own thread-local to store our
+// current thread's ID, and then we lazily initialize it to something allocated
+// from a global counter.
+#[cfg(target_feature = "atomics")]
+pub fn my_id() -> u32 {
+    use crate::sync::atomic::{AtomicU32, Ordering::SeqCst};
+
+    static NEXT_ID: AtomicU32 = AtomicU32::new(0);
+
+    #[thread_local]
+    static mut MY_ID: u32 = 0;
+
+    unsafe {
+        if MY_ID == 0 {
+            let mut cur = NEXT_ID.load(SeqCst);
+            MY_ID = loop {
+                if cur + 1 == 0 {
+                    crate::arch::wasm32::unreachable();
+                }
+                match NEXT_ID.compare_exchange(cur, cur + 1, SeqCst, SeqCst) {
+                    Ok(id) => break id + 1,
+                    Err(i) => cur = i,
+                }
+            };
         }
-        pub fn my_id() -> u32 {
-            unsafe { __wbindgen_current_id() }
-        }
-
-        // These are currently only ever used in `thread_local_atomics.rs`, if
-        // you'd like to use them be sure to update that and make sure everyone
-        // agrees what's what.
-        pub fn tcb_get() -> *mut u8 {
-            use crate::mem;
-            assert_eq!(mem::size_of::<*mut u8>(), mem::size_of::<u32>());
-            unsafe { __wbindgen_tcb_get() as *mut u8 }
-        }
-
-        pub fn tcb_set(ptr: *mut u8) {
-            unsafe { __wbindgen_tcb_set(ptr as u32); }
-        }
-
-        // FIXME: still need something for hooking exiting a thread to free
-        // data...
-
-    } else if #[cfg(target_feature = "atomics")] {
-        pub fn my_id() -> u32 {
-            panic!("thread ids not implemented on wasm with atomics yet")
-        }
-
-        pub fn tcb_get() -> *mut u8 {
-            panic!("thread local data not implemented on wasm with atomics yet")
-        }
-
-        pub fn tcb_set(_ptr: *mut u8) {
-            panic!("thread local data not implemented on wasm with atomics yet")
-        }
-    } else {
-        // stubbed out because no functions actually access these intrinsics
-        // unless atomics are enabled
+        MY_ID
```

Comment:
What is the reason for using a compare_exchange loop here, rather than an atomic increment (something like `MY_ID = NEXT_ID.fetch_add(1, SeqCst)`)? Is this not supported by wasm?

Reply:
While perhaps slightly overzealous, the thinking here is that if the counter everything should deterministically panic, as opposed to only one thread deterministically panicking but others continuing.

Reply:
I see. Maybe this is a nice opportunity to use the unstable `fetch_update` API, something like:

```
MY_ID = match NEXT_ID.fetch_update(|cur| cur.overflowing_add(1), SeqCst, SeqCst) {
    Some(id) => id + 1,
    Err(_) => crate::arch::wasm32::unreachable(),
}
```

Reply:
Oh good point! I think I'd prefer to avoid picking up an unstable feature for this since it's relatively minor, but I refactored a bit with your suggestion to use checked arithmetic which I think is better than what was there already!